### PR TITLE
重写重置流量提取逻辑

### PR DESCRIPTION
**Pull request by regendsoh**
- 提取shop内带重置套餐
- 匹配bought内用户最新购买的带重置套餐

NOTE: 先勿同意，等测试可行再审核
#TesterWelcome

### CODE REVIEW

**Comment by regendsoh** [comment]
https://github.com/Anankke/SSPanel-Uim/issues/1187

**Comment by Anankke** [comment]
还能比上一个拉垮？

**Comment by regendsoh** [comment]
@Anankke 俺布吉岛

**Comment by Colinwood11** [comment]
@Anankke Fatal error: Uncaught TypeError: Argument 1 passed to Illuminate\Database\Grammar::columnize() must be of the type array, string given, called in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php on line 121 and defined in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Grammar.php:107
Stack trace:
#0 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(121): Illuminate\Database\Grammar->columnize('id')
#1 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(75): Illuminate\Database\Query\Grammars\Grammar->compileColumns(Object(Illuminate\Database\Query\Builder), 'id')
#2 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(51): Illuminate\Database\Query\Grammars\Grammar->compileComponents(Object(Illuminate\Database\Query\Builder))
#3 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/MySqlGrammar.php(38): Illuminate\Database\Query\Grammars\Grammar->compileSelec in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Grammar.php on line 107
[root@HKSRV2483 ~]# php -n /www/wwwroot/webapi.backadd.tk/xcat Job CheckJob

Fatal error: Uncaught TypeError: Argument 1 passed to Illuminate\Database\Grammar::columnize() must be of the type array, string given, called in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php on line 121 and defined in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Grammar.php:107
Stack trace:
#0 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(121): Illuminate\Database\Grammar->columnize('id')
#1 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(75): Illuminate\Database\Query\Grammars\Grammar->compileColumns(Object(Illuminate\Database\Query\Builder), 'id')
#2 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/Grammar.php(51): Illuminate\Database\Query\Grammars\Grammar->compileComponents(Object(Illuminate\Database\Query\Builder))
#3 /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Query/Grammars/MySqlGrammar.php(38): Illuminate\Database\Query\Grammars\Grammar->compileSelec in /www/wwwroot/webapi.backadd.tk/vendor/illuminate/database/Grammar.php on line 107


我不确定可不可  但是我用toSql打出是这个错误

**Comment by Colinwood11** [comment]
Fatal error: Uncaught TypeError: Argument 1 passed to Illuminate\Database\Grammar::columnize() must be of the type array, string given, called in /www/wwwroot/vendor/illuminate/database/Query/Grammars/Grammar.php on line 121 and defined in /www/wwwroot/vendor/illuminate/database/Grammar.php:107
Stack trace:
#0 /www/wwwroot/vendor/illuminate/database/Query/Grammars/Grammar.php(121): Illuminate\Database\Grammar->columnize('id')
#1 /www/wwwroot/vendor/illuminate/database/Query/Grammars/Grammar.php(75): Illuminate\Database\Query\Grammars\Grammar->compileColumns(Object(Illuminate\Database\Query\Builder), 'id')
#2 /www/wwwroot/vendor/illuminate/database/Query/Grammars/Grammar.php(51): Illuminate\Database\Query\Grammars\Grammar->compileComponents(Object(Illuminate\Database\Query\Builder))
#3 /www/wwwroot/vendor/illuminate/database/Query/Grammars/MySqlGrammar.php(38): Illuminate\Database\Query\Grammars\Grammar->compileSelec in /www/wwwroot/vendor/illuminate/database/Grammar.php on line 107



垮了

**Comment by regendsoh** [comment]
@Colinwood11 $shopRenew = Shop::where('status','=','1')->where('content','like','%reset_value%')->where('content','like','%reset_exp%')->get('id')->toarray();
$boughts = Bought::whereIn('id', Bought::groupBy(userid)->where('shopid',$shopRenew->id)->max('id'));



试试看

**Comment by Colinwood11** [comment]
@regendsoh 刚通宵撸代码完  关机睡觉了

**Comment by regendsoh** [comment]
@Colinwood11 莫得问题 #TesterWelcome

**Comment by Colinwood11** [comment]
我开机了   问题依旧Argument 1 passed to Illuminate\Database\Grammar::columnize() must be of the type array, string given,

**Comment by Colinwood11** [comment]
Bought::groupBy('userid')    你又少了引号

**Comment by regendsoh** [comment]
`$shopRenew = Shop::where('status','1')->where('content','like','%reset_value%')->get(['id'])->toArray();
        $shopRenewId = Bought::whereIn('shopid',array_filter(array_column($shopRenew, 'id')))->groupBy('userid')->orderBy("id","desc")->get(['id']);
        $boughts = Bought::whereIn('id', array_filter(array_column(json_decode($shopRenewId), 'id')))->get();`

**Comment by Colinwood11** [comment]
@regendsoh 第一句话中   加了get(['id'])  的输出是null

**Comment by regendsoh** [comment]
image
@Colinwood11

**Comment by Anankke** [comment]
在？解决一下冲突？

**Comment by regendsoh** [comment]
在等大佬测试

> ```
> $shopRenew = Shop::where('status','1')->where('content','like','%reset_value%')->get(['id'])->toArray();
> $shopRenewId = Bought::whereIn('shopid',array_filter(array_column($shopRenew, 'id')))->groupBy('userid')->orderBy("id","desc")->get(['id']);
> $boughts = Bought::whereIn('id', array_filter(array_column(json_decode($shopRenewId), 'id')))->get();
> ```

**Comment by regendsoh** [comment]
image
@GeekQuerxy PR导致多了出来，先把我这三行测了ok才修改再merge

File: src/Command/Job.php [hunk 1]
```diff
@@ -141,8 +141,10 @@ public function DailyJob()
         }
 
         //auto reset
-        $boughts      = Bought::whereIn('shopid', $shopid)->get();
-        $bought_users = [];
+        $shopRenew = Shop::where('status','1')->where('content','like','%reset_value%')->get(['id'])->toArray();
```
Comment (CoreyAndres) [inline src/Command/Job.php:144]: 用 `->pluck('id')` 保护所有人的视力

File: src/Command/Job.php [hunk 2]
```diff
@@ -141,8 +141,10 @@ public function DailyJob()
...
+        $shopRenewId = Bought::whereIn('shopid',array_filter(array_column($shopRenew, 'id')))->groupBy('userid')->orderBy("id","desc")->get(['id']);
```
Comment (CoreyAndres) [inline src/Command/Job.php:145]: 同理。建议写法 `->select(new Expression('max("id") as id'))->pluck('id')`

**Pull request merged**